packages/backend/src/clients/cube/CubeClient.ts

```diff
@@ -133,20 +136,41 @@ export default class CubeClient implements SemanticLayerClient {
         query: SemanticLayerQuery,
         callback: (results: SemanticLayerResultRow[]) => void,
     ): Promise<number> {
+        const queryLimit = query.limit ?? this.maxQueryLimit;
+        let partialResultsLimit = Math.min(100, queryLimit);
+        let prevLimit = partialResultsLimit;
         let offset = 0;
-        const limit = 100;
         let partialResults: SemanticLayerResultRow[] = [];
+
         do {
             /* eslint-disable-next-line no-await-in-loop */
             partialResults = await this.getResults({
                 ...query,
                 offset,
-                limit,
+                limit: partialResultsLimit,
             });
+
             callback(partialResults);
-            offset += limit;
-        } while (partialResults.length === limit);
-        return offset + partialResults.length;
+
+            // update the offset
+            offset += partialResults.length;
+
+            // keep track of the previous limit so that we can stop when result count is less than it
+            prevLimit = partialResultsLimit;
+
+            // decrease the limit if we are close to the query limit
+            partialResultsLimit = Math.min(
+                partialResultsLimit,
+                queryLimit - offset,
+            );
+        } while (
+            partialResults.length > 0 &&
+            partialResults.length >= prevLimit &&
+            offset < queryLimit
```

Comment:
`queryLimit` will always be either the limit set by the query itself, or the value of `lightdashConfig.query.maxLimit` which has a default value of `5000`: `const queryLimit = query.limit ?? this.maxQueryLimit`

The loop stops running when:
- There are no results
- The number of results are less than the partial results limit used to fetch them `prevLimit` - if there are less results it means that we reached the end
- The `offset` is bigger or equal to the whole `queryLimit` which means we should stop querying because we reached the limit of results we wanted

Also, important to note, the `partialResultsLimit` is the amount of results we fetch on each loop and gets initialized as the minimum between 100 or the query limit in the case that we want to fetch less than 100 results.
During the loop we update it to be the minimum between the partial results limit or the remaining results left to fetch (`queryLimit - offset`), this way we won't over fetch results

Reply:
should I update some of these variable names to make it a bit more easier to follow?
e.g.

- `partialResultsLimit` -> `partialResultsLimit`
- `prevLimit` -> `previousPartialResultsLimit`
- set 100 as `this.maxPartialResultsLimit`